digdag-core/src/main/java/io/digdag/core/agent/GrantedPrivilegedVariables.java

```diff
@@ -0,0 +1,152 @@
+import com.fasterxml.jackson.databind.JsonNode;
+import com.google.common.base.Optional;
+import com.google.common.collect.ImmutableList;
+import io.digdag.client.config.Config;
+import io.digdag.client.config.ConfigKey;
+import io.digdag.client.config.ConfigException;
+import io.digdag.spi.PrivilegedVariables;
+import io.digdag.spi.SecretAccessContext;
+import io.digdag.spi.SecretAccessPolicy;
+import io.digdag.spi.SecretNotFoundException;
+import io.digdag.spi.SecretProvider;
+import io.digdag.spi.SecretScopes;
+import io.digdag.spi.SecretStore;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Supplier;
+
+public class GrantedPrivilegedVariables
+    implements PrivilegedVariables
+{
+    static SecretProvider privilegedSecretProvider(SecretAccessContext context, SecretAccessPolicy accessPolicy, SecretStore store)
+    {
+        return (key) -> {
+            if (!accessPolicy.isSecretAccessible(context, key)) {
+                return Optional.absent();
+            }
+
+            Optional<String> projectSecret = store.getSecret(context.projectId(), SecretScopes.PROJECT, key);
+
+            if (projectSecret.isPresent()) {
+                return projectSecret;
+            }
+
+            return store.getSecret(context.projectId(), SecretScopes.PROJECT_DEFAULT, key);
+        };
+    }
+
+    private static class SecretOnlyGrant
+    {
+        @JsonProperty
+        ConfigKey secret;
+    }
+
+    private interface VariableAccessor
+    {
+        String get(boolean required);
+    }
+
+    public static GrantedPrivilegedVariables empty()
+    {
+        return new GrantedPrivilegedVariables(new LinkedHashMap<>());
+    }
+
+    public static GrantedPrivilegedVariables build(
+            Config grants,
+            Config params,
+            SecretProvider secretProvider)
+    {
+        Map<String, VariableAccessor> variables = new LinkedHashMap<>();
+        for (String key : grants.getKeys()) {
+            variables.put(key, buildAccessor(grants, key, params, secretProvider));
+        }
+        return new GrantedPrivilegedVariables(variables);
+    }
+
+    private static VariableAccessor buildAccessor(
+            Config grants, String key,
+            Config params,
+            SecretProvider secretProvider)
+    {
+        if (grants.get(key, JsonNode.class).isObject()) {
+            ConfigKey secretOnlyKey = grants.getNested(key).convert(SecretOnlyGrant.class).secret;
+            return (required) -> {
+                Optional<String> secret = secretProvider.getSecretOptional(secretOnlyKey.toString());
+                if (required && !secret.isPresent()) {
+                    throw new SecretNotFoundException(secretOnlyKey.toString());
+                }
+                return secret.orNull();
+            };
+        }
+        else {
+            ConfigKey secretSharedKey = grants.get(key, ConfigKey.class);
```

Comment:
How about simply saying "secret access" and "secret-only access"? Public method names are:

* ConfigSelector.Builder.addSecretAccess(String...)
* ConfigSelector.Builder.addSecretOnlyAccess(String...)
* ConfigSelector.Builder.addAllSecretAccess(Iterable<? extends String>)
* ConfigSelector.Builder.addAllSecretOnlyAccess(Iterable<? extends String>)

Local variables can be like this:

```
+            ConfigKey secretAccessKey = grants.get(key, ConfigKey.class);
+            return (required) -> {
 +                Optional<String> secret = secretProvider.getSecretOptional(secretAccessKey.toString());
...
+                String value = nested.get(secretAccessKey.getLastName(), String.class, null);
```